packages/wp-source/src/actions.ts

```diff
@@ -105,6 +145,48 @@ const actions: WpSource["actions"]["source"] = {
     } catch (e) {
       // It's a server error (4xx or 5xx).
       if (e instanceof ServerError) {
+        // Check if the error is 4xx and we want to handle redirections in case
+        // of an error.
+        if (e.status === 404) {
+          // If the route matched the redirection regex or
+          // state.router.redirections === 'all', then we have already started
+          // fetching the data, so just await it.
+          let head = redirectionPromise && (await redirectionPromise);
+
+          const { redirections } = state.router;
+          if (
+            !head &&
+            (redirections === "404" ||
+              (Array.isArray(redirections) && redirections.includes("404")))
+          ) {
+            try {
+              head = await fetch(state.source.url + link, { method: "HEAD" });
+            } catch (e) {
+              // If there is no redirection, we ignore it and just continue
+              // handling the 404 ServerError that was thrown previously.
+            }
+          }
+
+          if (head?.redirected) {
+            const pathname = new URL(head.url).pathname;
+
+            const searchParams = stringify(query);
+
+            const location = `${normalize(pathname)}${
+              searchParams && "?"
+            }${searchParams}`;
```

Comment:
My thinking was that the redirections should match some pathname like `/url/some-post` and always redirect for that pathname regardless of the query or hash. But you are right that a user can set up a redirection which includes a query string or hash as well like: `/url/some-post?key=value` --> `/url/some-new-post`. We have to add code to account for that use case.

But I still think that we should preserve the query and hash when doing a redirection if there is no query or hash present in the redirected link. Otherwise, the user would have to set up a redirection for every possible query and hash that can be added to the link and I think that might not always be practical.

So, I think that we have 3 cases here:

1. `/url/some-post` --> `/url/redirected` (no query or hash)
  
   In this case, we should preserve the hash and query if they are present in the link, because and redirect the user. So, `/url/some-post?key=value` redirects to `/url/redirected?key=value`.

2. `/url/some-post?key=value` --> `/url/redirected` (original link has a query)

   If the original link includes query or hash, we have to extract it and omit it from the query params in the redirected link. So, `/url/some-post?key=value&p=123` redirects to `/url/redirected?p=123`.

3. `/url/some-post` --> `/url/redirected?key=value` (The redirected link has hash or query)

   This is simple and similar to case 1: We add whatever query and hash is present on the link when accessing it, just making sure to use the `&` instead of `?` when glueing together the final link.

Reply:
I am sorry but I don't think that is right Michal. We should not mess with the `Location` that is returned by the server. If they want to preserve all the queries, part of the queries, or none of the queries, that is up to the redirection logic.

Take into account that in the "server-side rendered world" there is no "extra step" like we have here to modify the redirection. So whatever is returned by the `Location` header is used by the browser.

Redirections, when set properly, take into account the queries and pass them to the next link:

- Apache: `RedirectMatch ^/images/(.*)$ http://images.example.com/$1`
- NGINX: `rewrite ^/images/(.*)$ http://docs.domain2.com/$1 redirect;`

But this the standard, they can be way more complex. This one changes the value of `some-query` when it is `some-value`:

```nginx
location /some-path {
   if ($args ~* "(.*)some-query=some-value(.*)") {
       set $args "$1some-query=other-value$2";
       return 301 $scheme://$host$uri$args;
   }

}
```

Vercel and Netlify have simpler redirection setups where they pass the queries by default, but Netlify also has a more complex way to define the redirections where you can modify the queries: https://docs.netlify.com/routing/redirects/redirect-options/#query-parameters

The Redirections plugin has options for what to do with the queries as well. And of course, you can also use RegExps.

<img width="527" alt="Screenshot 2020-11-19 at 08 59 37" src="https://user-images.githubusercontent.com/3305402/99637781-9ff9f280-2a45-11eb-93cf-f9c13850c820.png">

So I really think we should respect the `Location` header (with the exception of the `frontity_` queries in the server, of course).